src/libcryptobox/cryptobox.c

```diff
@@ -677,6 +673,59 @@ void rspamd_cryptobox_sign(unsigned char *sig, unsigned long long *siglen_p,
 	}
 }
 
+#if OPENSSL_VERSION_MAJOR >= 3
+bool rspamd_cryptobox_verify_compat(const unsigned char *sig,
+									gsize siglen,
+									const unsigned char *m,
+									gsize mlen,
+									const rspamd_pk_t pk,
+									const rspamd_sk_t sk,
+									enum rspamd_cryptobox_mode mode) {
+	bool ret = false;
+
+	if (G_LIKELY(mode == RSPAMD_CRYPTOBOX_MODE_25519)) {
+		if (siglen == rspamd_cryptobox_signature_bytes(RSPAMD_CRYPTOBOX_MODE_25519)) {
+			ret = (crypto_sign_verify_detached(sig, m, mlen, sk) == 0);
+		}
+	}
+	else {
+#ifndef HAVE_USABLE_OPENSSL
+		g_assert(0);
+#else
+		EVP_MD_CTX *sha_ctx;
+		unsigned char h[64];
+
+		/* Prehash */
+		sha_ctx = EVP_MD_CTX_create();
+		g_assert(EVP_DigestInit(sha_ctx, EVP_sha512()) == 1);
+		EVP_DigestUpdate(sha_ctx, m, mlen);
+		EVP_DigestFinal(sha_ctx, h, NULL);
+
+
+		OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
+		EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
```

Comment:
You should not produce public key from private, you need to use public key for verifications - that's how digital signatures ever works.